### ISSUE TITLE
[MNG-6490] Maven shall not fail reporting circular dependency when the dependency is a classified secondary artifact

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7375]
**[John Canny](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jfcanny)** opened **[MNG-6490](https://issues.apache.org/jira/browse/MNG-6490?redirect=false)** and commented

As of maven 3.5.2, 3.5.3, 3.5.4, the following pom fails with the error

"dependencies.dependency. Main:MainJar:1' for Main:MainJar:1 is referencing itself"

But the dependency is not circular, it references a classified jar (in our use cases its an architecture-dependent native code container jar). The pom below allows the main jar to be built without building the dependency every time (other lines conditionally build the dependency), and ensures the appropriate pre-built dependency is loaded. Behavior in maven 3.5.0 and earlier was correct (i.e. no error). This breaks several of the usage scenarios for classified artifacts...

 

```xml
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0
                             http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>Main</groupId>
  <artifactId>MainJar</artifactId>
  <packaging>jar</packaging>
  <version>1</version>
  <dependencies>
    <dependency>
      <groupId>${project.groupId}</groupId>
      <artifactId>${project.artifactId}</artifactId>
      <version>${project.version}</version>
      <classifier>linux</classifier>
    </dependency>
  </dependencies>
</project>
```

---

**Affects:** 3.5.2, 3.5.3, 3.5.4

**Issue Links:**
- [MNG-6541](https://issues.apache.org/jira/browse/MNG-6541) less restrictive "is referencing itself" check
 (_**"is duplicated by"**_)
- [MNG-6123](https://issues.apache.org/jira/browse/MNG-6123) detect self references in POM and fail fast
 (_**"is caused by"**_)

**Remote Links:**
- [GitHub Pull Request #188
](https://github.com/apache/maven/pull/188)

0 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I'd even add this fix to 3.6.0.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Done in [e4b53d3f2c00e00d78129e20e177399a12d27847](https://gitbox.apache.org/repos/asf?p=maven.git;a=commitdiff;h=e4b53d3f2c00e00d78129e20e177399a12d27847)
